regparser/web/jobs/views.py

```diff
@@ -148,3 +194,172 @@ def retrieve(self, request, *args, **kwargs):
+    lookup_field = "hexhash"
+    size_limit = 100000000  # Arbitrary 100MB limit.
+
+    def create(self, request, *args, **kwargs):
+        """
+        Overrides the ``create`` method of ``mixins.CreateModelMixin`` in order
+        to add the file contents to the database.
+
+        Side effects
+            Alters the DB.
+
+        :arg HttpRequest request: the incoming request.
+
+        :rtype: Response
+        :returns: JSON or HTML of the information about the file (status 201),
+            or about why the file couldn't be added (status 400).
+        """
+        serialized = self.get_serializer(data=request.data)
+        serialized.is_valid(raise_exception=True)
+
+        uploaded_file = request.data["file"]
+        if uploaded_file.size > self.size_limit:
+            return Response(dict(error="File too large (%s-byte limit)." %
+                                 self.size_limit),
+                            status=status.HTTP_400_BAD_REQUEST)
+        if uploaded_file.multiple_chunks():
+            contents = b"".join(chunk for chunk in uploaded_file.chunks())
+        else:
+            contents = uploaded_file.read()
+        md = hashlib.md5(contents)
+        hexhash = md.hexdigest()
+        filename = uploaded_file.name
+        url = file_url(hexhash)
+
+        if RegulationFile.objects.filter(hexhash=hexhash).count() > 0:
+            return Response(dict(error="File already present."),
+                            status=status.HTTP_400_BAD_REQUEST)
+        else:
+            serialized.save(contents=contents, file=uploaded_file,
+                            filename=filename, hexhash=hexhash, url=url)
+
+        headers = self.get_success_headers(serialized.data)
+        return Response(serialized.data, status=status.HTTP_201_CREATED,
+                        headers=headers)
+
+    def post(self, request, *args, **kwargs):
+        return self.create(request, *args, **kwargs)
+
+    def get(self, request, *args, **kwargs):
+        return self.list(request, *args, **kwargs)
+
+
+class FileUploadViewInstance(mixins.RetrieveModelMixin,
+                             mixins.UpdateModelMixin, mixins.DestroyModelMixin,
+                             generics.GenericAPIView):
+    serializer_class = FileUploadSerializer
+    queryset = RegulationFile.objects.all()
+    lookup_field = "hexhash"
+
+    def get(self, request, *args, **kwargs):
+        """
+        Overrides the method from ``RetrieveModelMixin`` so that we return the
+        contents of the file instead of a JSON object representing the file.
+
+        Impure
+            Reads from the DB.
+
+        :arg HttpRequest request: the incoming request.
+
+        :rtype: Response
+        :returns: The raw contents of the file.
+        """
+        # Is the next line the best way to kick off a 404 if there's no match?
+        self.retrieve(request, *args, **kwargs)
+
+        uploaded_file = RegulationFile.objects.get(hexhash=kwargs["hexhash"])
```

Comment:
Technically, yes, but no, because of the `self.retrieve()` line above it, which will short-circuit and kick back a 404 response. Including the `self.retrieve()` line seems like the simplest and shortest way to accomplish this, as otherwise I think a bunch of exception-handling and writing out what to pass back as a 404 is necessary. I decided to use `self.retrieve()` for this purpose in order to take advantage of the framework, but I'm not 100% certain this is correct.